src/SignalR/common/Shared/AsyncEnumerableAdapters.cs

```diff
@@ -13,38 +11,57 @@ namespace Microsoft.AspNetCore.SignalR.Internal
+        public static IAsyncEnumerator<object?> MakeCancelableAsyncEnumerator<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationToken cancellationToken = default)
         {
-            return new CancelableAsyncEnumerable<T>(asyncEnumerable, cancellationToken);
+            var enumerator = asyncEnumerable.GetAsyncEnumerator(cancellationToken);
+            return enumerator as IAsyncEnumerator<object?> ?? new BoxedAsyncEnumerator<T>(enumerator);
         }
 
         public static IAsyncEnumerable<T> MakeCancelableTypedAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationTokenSource cts)
         {
             return new CancelableTypedAsyncEnumerable<T>(asyncEnumerable, cts);
         }
 
-#if NETCOREAPP
-        public static async IAsyncEnumerable<object?> MakeAsyncEnumerableFromChannel<T>(ChannelReader<T> channel, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        public static IAsyncEnumerator<object?> MakeAsyncEnumeratorFromChannel<T>(ChannelReader<T> channel, CancellationToken cancellationToken = default)
         {
-            await foreach (var item in channel.ReadAllAsync(cancellationToken))
-            {
-                yield return item;
-            }
+            return new ChannelAsyncEnumerator<T>(channel, cancellationToken);
         }
-#else
-        // System.Threading.Channels.ReadAllAsync() is not available on netstandard2.0 and netstandard2.1
-        // But this is the exact same code that it uses
-        public static async IAsyncEnumerable<object?> MakeAsyncEnumerableFromChannel<T>(ChannelReader<T> channel, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+
+        private class ChannelAsyncEnumerator<T> : IAsyncEnumerator<object?>
         {
-            while (await channel.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            private readonly ChannelReader<T> _channel;
+            private readonly CancellationToken _cancellationToken;
+            public ChannelAsyncEnumerator(ChannelReader<T> channel, CancellationToken cancellationToken)
+            {
+                _channel = channel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public object? Current { get; private set; }
+
+            public ValueTask<bool> MoveNextAsync()
             {
-                while (channel.TryRead(out var item))
+                if (_channel.TryRead(out var item))
                 {
-                    yield return item;
+                    Current = item;
+                    return new ValueTask<bool>(true);
                 }
+
+                return new ValueTask<bool>(MoveNextAsyncAwaited());
             }
+
+            private async Task<bool> MoveNextAsyncAwaited()
+            {
+                if (await _channel.WaitToReadAsync(_cancellationToken) && _channel.TryRead(out var item))
```

Comment:
I did want to keep the `.ConfigureAwait(false)` in client code (and code in the `/Shared/` folder that could be used by the client).

```suggestion
                if (await _channel.WaitToReadAsync(_cancellationToken).ConfigureAwait(false) && _channel.TryRead(out var item))
```

Reply:
OK, I was actually wondering on that, makes sense now.

Reply:
It was easy enough to commit from the github UI, so I did it for you. Thanks again!